pkg/kubectl/cmd/apply.go

```diff
@@ -191,11 +233,112 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
 	if err != nil {
 		return err
 	}
-
 	if count == 0 {
 		return fmt.Errorf("no objects passed to apply")
 	}
 
+	if !options.Prune {
+		return nil
+	}
+
+	selector, err := labels.Parse(options.Selector)
+	if err != nil {
+		return err
+	}
+	p := pruner{
+		mapper:        mapper,
+		clientFunc:    f.ClientForMapping,
+		clientsetFunc: f.ClientSet,
+
+		selector:    selector,
+		visitedUids: visitedUids,
+
+		cascade:     options.Cascade,
+		gracePeriod: options.GracePeriod,
+
+		out: out,
+	}
+	for n, _ := range visitedNamespaces {
+		for _, m := range visitedNamespacedRESTMappings {
+			if err := p.prune(n, m, shortOutput); err != nil {
+				return fmt.Errorf("error pruning objects: %v", err)
+			}
+		}
+	}
+	for _, m := range visitedNonNamespacedRESTMappings {
+		if err := p.prune(api.NamespaceNone, m, shortOutput); err != nil {
+			return fmt.Errorf("error pruning objects: %v", err)
+		}
+	}
+
+	return nil
+}
+
+type pruner struct {
```

Comment:
To me it'd be perfectly reasonable to put have this in a separate file, when I was trying to read `apply.go` last time, it seemed rather big and not quite structured.